File: solutions/4086-lexicographically-smallest-string-after-deleting-duplicate-characters/lexicographically-smallest-string-after-deleting-duplicate-characters.java
class Solution {
    public String lexSmallestAfterDeletion(String s) {
        char[] a = s.toCharArray();
        int[] rem = new int[26];
        for (char c : a) rem[c - 'a']++;

        int[] cnt = new int[26];
        StringBuilder sb = new StringBuilder();

        for (char c : a) {
            rem[c - 'a']--;
            while (sb.length() > 0) {
                char t = sb.charAt(sb.length() - 1);
                if (t > c && (rem[t - 'a'] > 0 || cnt[t - 'a'] > 1)) {
                    cnt[t - 'a']--;
                    sb.setLength(sb.length() - 1);
                } else break;
            }
            sb.append(c);
            cnt[c - 'a']++;
        }

        while (sb.length() > 0) {
            int i = sb.length() - 1;
            char c = sb.charAt(i);
            if (cnt[c - 'a'] > 1) {
                cnt[c - 'a']--;
                sb.setLength(i);
            } else break;
        }

        return sb.toString();
    }
}

File: solutions/4086-lexicographically-smallest-string-after-deleting-duplicate-characters/lexicographically-smallest-string-after-deleting-duplicate-characters.cpp
class Solution {
public:
    string lexSmallestAfterDeletion(string s) {
        int r[26] = {0}, in[26] = {0};
        for (char c : s) r[c - 'a']++;
        string ans;
        for (char c : s) {
            int i = c - 'a';
            r[i]--;
            while (!ans.empty()) {
                int j = ans.back() - 'a';
                if (ans.back() > c && (r[j] > 0 || in[j] > 1)) {
                    in[j]--;
                    ans.pop_back();
                } else break;
            }
            ans += c;
            in[i]++;
        }
        while (!ans.empty() && in[ans.back() - 'a'] > 1) {
            in[ans.back() - 'a']--;
            ans.pop_back();
        }
        return ans;
    }
};